README.md
<!--
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-24 17:20:53
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-09-24 17:22:42
 * @FilePath: /phoenix_rating/README.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

phoenix 将做为企业级基础组件:Rating,提供项目支持。

## Features

phoenix 将做为企业级基础组件:Rating,提供项目支持。

## Getting started

phoenix 将做为企业级基础组件:Rating,提供项目支持。

## Usage

phoenix 将做为企业级基础组件:Rating,提供项目支持。

```dart
const like = 'sample';
```

## Additional information

phoenix 将做为企业级基础组件:Rating,提供项目支持。


CHANGELOG.md
<!--
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-24 17:20:53
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 23:14:31
 * @FilePath: /phoenix_rating/CHANGELOG.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
## 1.0.0

* [新建] 企业级基础组件:Rating

## 1.0.1

* [更新] resource path

## 1.0.2

* [新增] example

test/phoenix_rating_test.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2023-09-24 17:20:53
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-09-24 17:25:58
 * @FilePath: /phoenix_rating/test/phoenix_rating_test.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import 'package:flutter_test/flutter_test.dart';

import 'package:phoenix_rating/phoenix_rating.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/rating_star.dart
import 'package:flutter/material.dart';
import 'package:phoenix_base/phoenix.dart';

import 'extension/rating_assets.dart';

/// 星星视图的自定义构造器
/// state，[RatingState] 星星状态
typedef RatingStarBuilder = Widget Function(RatingState state);

/// 星星状态
enum RatingState {
  /// 半颗
  half,

  /// 未选
  unselect,

  /// 已选
  select,
}

/// 星级评分控件，支持：
///
/// * 可自定义图片、颜色、大小、间距
/// * 支持点击选中
/// * 支持是否限制评分最少一颗星，即第一颗星支持是否可反选
/// * 支持半颗星（仅支持展示，不支持选择）
class RatingStar extends StatefulWidget {
  static const DEFAULT_COUNT = 5;
  static const DEFAULT_SPACE = 1.0;

  /// 星星的总数，默认为 5 颗
  final int count;

  /// 初始选中个数
  final double selectedCount;

  /// 星星间的水平间距，默认为 1.0
  final double space;

  /// 是否可评 0 颗星，即第一颗星是否支持反选，默认不可评 0 星
  final bool canRatingZero;

  /// 单颗星星视图的自定义构造器
  final RatingStarBuilder? starBuilder;

  /// 如果设置了，就支持编辑
  final ValueChanged<int>? onSelected;

  const RatingStar({
    Key? key,
    this.count = DEFAULT_COUNT,
    this.selectedCount = 0,
    this.space = DEFAULT_SPACE,
    this.starBuilder,
    this.onSelected,
    this.canRatingZero = false,
  }) : super(key: key);

  @override
  _BrnRatingStarState createState() => _BrnRatingStarState();
}

class _BrnRatingStarState extends State<RatingStar> {
  late double currSelected;

  @override
  void initState() {
    super.initState();
    currSelected = widget.selectedCount;
  }

  @override
  void didUpdateWidget(RatingStar oldWidget) {
    currSelected = widget.selectedCount;
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.start,
      children: _getContent(),
    );
  }

  List<Widget> _getContent() {
    List<Widget> list = [];
    for (var i = 0; i < widget.count; i++) {
      RatingState state;
      if (i < currSelected.floor()) {
        state = RatingState.select;
      } else if (i == currSelected.floor() && i < currSelected.ceil()) {
        state = RatingState.half;
      } else {
        state = RatingState.unselect;
      }
      var rating = widget.starBuilder != null
          ? widget.starBuilder!(state)
          : _buildRating(state);

      if (widget.onSelected != null) {
        list.add(GestureDetector(
          child: rating,
          onTap: () {
            // 反选第一个
            if (i == 0 && currSelected == 1 && widget.canRatingZero) {
              currSelected = 0;
            } else {
              currSelected = (i + 1).toDouble();
            }
            widget.onSelected!(currSelected.toInt());
            setState(() {});
          },
          behavior: HitTestBehavior.opaque,
        ));
      } else {
        list.add(rating);
      }

      if (i != widget.count - 1) {
        list.add(SizedBox(
          width: widget.space,
          height: 1,
        ));
      }
    }
    return list;
  }

  Widget _buildRating(RatingState state) {
    switch (state) {
      case RatingState.select:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStar, 16, 16,
            package: 'phoenix_rating');
      case RatingState.half:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStarHalf, 16, 16,
            package: 'phoenix_rating');
      case RatingState.unselect:
      default:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStar, 16, 16,
            color: const Color(0xFFF0F0F0), package: 'phoenix_rating');
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'rating_example.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: RatingExample(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


example/lib/rating_example.dart
/*
 * @Author: lipeng 1162423147@qq.com
 * @Date: 2022-04-29 17:06:50
 * @LastEditors: lipeng 1162423147@qq.com
 * @LastEditTime: 2023-10-12 23:09:17
 * @FilePath: /phoenix_rating/example/lib/rating_example.dart
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AEap
 */
import 'package:flutter/material.dart';
import 'package:phoenix_base/phoenix.dart';
import 'package:phoenix_rating/extension/rating_assets.dart';
import 'package:phoenix_rating/phoenix_rating.dart';

/// 星级评分条
class RatingExample extends StatefulWidget {
  @override
  _RatingExampleState createState() => _RatingExampleState();
}

class _RatingExampleState extends State<RatingExample> {
  var num = 3;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('星级评分控件Example'),
      ),
      body: Padding(
        padding: const EdgeInsets.all(20.0),
        child: ListView(
          children: <Widget>[
            // 只接受整数，外界
            const Text("支持半颗"),
            const RatingStar(),
            const RatingStar(
              selectedCount: 0.5,
            ),
            const RatingStar(
              selectedCount: 3.1,
            ),
            const RatingStar(
              selectedCount: 3.6,
              count: 10,
            ),
            const SizedBox(height: 20),
            GestureDetector(
              child: const Text("支持点击选中，第一个支持反选"),
              onTap: () {
                // BrnToast.show("haha", context);
                setState(() {
                  num = 4;
                });
              },
            ),
            RatingStar(
              selectedCount: num.toDouble(),
              space: 5,
              canRatingZero: true,
              onSelected: (count) {
                // BrnToast.show("选中了$count个", context);
              },
            ),
            const SizedBox(height: 20),
            const Text("自定义图片，颜色，大小"),
            RatingStar(
              selectedCount: 3,
              space: 1,
              canRatingZero: true,
              onSelected: (count) {
                // BrnToast.show("选中了$count个", context);
              },
              starBuilder: _buildRating,
            )
          ],
        ),
      ),
    );
  }

  // 自定义图片，大小，颜色
  Widget _buildRating(RatingState state) {
    switch (state) {
      case RatingState.select:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStar, 16, 16,
            color: const Color(0xFF3571DC), package: 'phoenix_rating');
      case RatingState.half:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStarHalf, 16, 16,
            package: 'phoenix_rating');
      case RatingState.unselect:
      default:
        return PhoenixTools.getAssetSizeImage(RatingAssets.iconStar, 16, 16,
            color: const Color(0xFFF0F0F0), package: 'phoenix_rating');
    }
  }
}
